### PR TITLE
`grammars`: cache decoded token codepoints for faster sampling

### PR DESCRIPTION
**Pull request by ochafik**
**Edit**: superseded by https://github.com/ggerganov/llama.cpp/pull/7587 & https://github.com/ggerganov/llama.cpp/pull/7424, can't measure significant improvements from caching the decoded codepoints: closing this PR.

Cache token pieces & their decoded codepoints (for common case where there's no partial utf8 sequence prefix) in `llama_context` to speed up grammar sampling (https://github.com/ggerganov/llama.cpp/issues/4218)

cc/ @ejones, @HanClinto

**Update**: while this PR still seems useful, https://github.com/ggerganov/llama.cpp/pull/7424 is bringing the most performance in most cases. I've run [some benchmarks](https://gist.github.com/ochafik/56733ecd349874fd66cedcfdfa619e9c) that seems to show this PR will add some marginal gains upon it.

This simple example runs **1.9x faster** (total time; __sampling itself__ goes from ~16 ms per token to ~1.9 ms per token, a **9x sampling speedup**):

```bash
# git remote add ochafik https://github.com/ochafik/llama.cpp.git
# git fetch ochafik
hyperfine --warmup 1 --runs 5 \
    -L branch ochafik/grammar-fast,master \
    --setup 'git checkout {branch} && make clean && make -j LLAMA_CURL=1 main' \
    'BRANCH={branch} \
        ./main --seed 12345 \
            -mu https://huggingface.co/TheBloke/phi-2-GGUF/resolve/main/phi-2.Q4_K_M.gguf \
            -j '"'"'{"items": {"type": "number"}, "minItems": 10, "maxItems": 100}'"'"' \
            -p "JSON list of first 50 integers"'
```

<details>
<summary>Show output</summary>

```bash
Benchmark 1: BRANCH=grammar-fast \
        ./main --seed 12345 \
            -m models/phi-2.Q4_K_M.gguf \
            -j '{"items": {"type": "number"}, "minItems": 10, "maxItems": 100}' \
            -p "JSON list of first 50 integers"
  Time (mean ± σ):      3.229 s ±  0.039 s    [User: 0.621 s, System: 0.191 s]
  Range (min … max):    3.193 s …  3.290 s    5 runs
 
Benchmark 2: BRANCH=master \
        ./main --seed 12345 \
            -m models/phi-2.Q4_K_M.gguf \
            -j '{"items": {"type": "number"}, "minItems": 10, "maxItems": 100}' \
            -p "JSON list of first 50 integers"
  Time (mean ± σ):      6.398 s ±  0.073 s    [User: 3.629 s, System: 0.258 s]
  Range (min … max):    6.345 s …  6.523 s    5 runs
 
Summary
  'BRANCH=grammar-fast \
        ./main --seed 12345 \
            -m models/phi-2.Q4_K_M.gguf \
            -j '{"items": {"type": "number"}, "minItems": 10, "maxItems": 100}' \
            -p "JSON list of first 50 integers"' ran
    1.98 ± 0.03 times faster than 'BRANCH=master \
        ./main --seed 12345 \
            -m models/phi-2.Q4_K_M.gguf \
            -j '{"items": {"type": "number"}, "minItems": 10, "maxItems": 100}' \
            -p "JSON list of first 50 integers"'
```

`master`:

```
llama_print_timings:        load time =    2416.78 ms
llama_print_timings:      sample time =    3274.38 ms /   203 runs   (   16.13 ms per token,    62.00 tokens per second)
llama_print_timings: prompt eval time =      49.09 ms /     7 tokens (    7.01 ms per token,   142.59 tokens per second)
llama_print_timings:        eval time =    2450.08 ms /   202 runs   (   12.13 ms per token,    82.45 tokens per second)
llama_print_timings:       total time =    6037.77 ms /   209 tokens
```

this PR:

```
llama_print_timings:        load time =     183.22 ms
llama_print_timings:      sample time =     377.09 ms /   203 runs   (    1.86 ms per token,   538.33 tokens per second)
llama_print_timings: prompt eval time =      49.03 ms /     7 tokens (    7.00 ms per token,   142.77 tokens per second)
llama_print_timings:        eval time =    2411.72 ms /   202 runs   (   11.94 ms per token,    83.76 tokens per second)
llama_print_timings:       total time =    2887.39 ms /   209 tokens
```

</details>

Note: removed the following changes from earlier version of this PR
- https://github.com/ggerganov/llama.cpp/pull/7370)
- Non-quadratic `llama_grammar_copy` for speculative decoding (will send this separately)

### CODE REVIEW

File: llama.cpp [hunk 1]
```diff
@@ -12727,6 +12727,10 @@ static std::vector<llama_grammar_candidate> llama_grammar_reject_candidates_for_
         }
     }
 
+    if (next_candidates.empty()) {
+        return rejects;
```
Comment (HanClinto) [inline llama.cpp:12731]: Such an elegant optimization. Very nice catch!

**Comment by HanClinto** [comment]
I've been out of it for a few days, but finally digging back in. GREAT work on all of this, @ochafik !!

> Cache token pieces & their decoded codepoints in `llama_sample_grammar`

I'm still trying to read through and understand this -- I can't claim to have a good handle on this yet. Are you able to summarize why this optimization works?

> Non-quadratic `llama_grammar_copy` for speculative decoding

Because `llama_grammar_copy` is only used by `speculative.cpp`, I wonder if this change should be left out for now...? Unless the changes to retaining cached codepoints makes it easier to keep it in here, then let's keep it in here for now. Instead of profiling on speculative.cpp, might be easiest to write unit tests around this?

**Comment by ochafik** [comment]
>> Non-quadratic llama_grammar_copy for speculative decoding

>Because llama_grammar_copy is only used by speculative.cpp, I wonder if this change should be left out for now...?

Good point, reverted / will send separately 👌 (one less thing to check for this PR :-D)

File: llama.cpp (hunk 1 above)
Comment (ochafik) [reply llama.cpp:12731]: Took me forever (and many failed refactorings) to understand what the code was doing tbh 😅

**Comment by ochafik** [comment]
> > Cache token pieces & their decoded codepoints in llama_sample_grammar

> I'm still trying to read through and understand this -- I can't claim to have a good handle on this yet. Are you able to summarize why this optimization works?

The code used to decode each token in the context of a potentially incomplete [UTF-8](https://en.wikipedia.org/wiki/UTF-8#Encoding) sequence (say, if the first token was a byte token with the `110xxxxx` bit mask). This is very rare in practice, so tokens can be generally decoded to unicode codepoints w/o any partial utf bytes as prefix. I'm caching said non-partial decoding, which saves a lot of time (profiling the code showed lots of time was spent here). Also caching the token piece although that bit is arguably less useful (only in case of partial utf decoding, maybe more frequent for non-latin languages?).

**Comment by HanClinto** [comment]
> > I've been out of it for a few days, but finally digging back in
> 
> @HanClinto welcome back, hope the robotics competition went well!

Thanks -- it went really well! The kids all did great! 10 kids with 9 robots in 3 different categories. We took home two special nominations, two gold trophies, a silver, and a bronze. Tried a new thing this year where our team jerseys were custom-printed lab coats for all of the kids, and that was a ton of fun. :D It's a pretty big competition (here's a [highlight reel from last year](https://www.youtube.com/watch?v=Xj-u3SVRpfI)), and the kids are excited for next year.

Anyways, enough fluff. That's what happens when a project doesn't have a Discord server -- we clutter up the PRs with such conversations. :D

> > > Non-quadratic llama_grammar_copy for speculative decoding
> 
> > Because llama_grammar_copy is only used by speculative.cpp, I wonder if this change should be left out for now...?
> 
> Good point, reverted / will send separately 👌 (one less thing to check for this PR :-D)

Sounds great! 👍 

I know I've mentioned it a few times, but I have a rather complex SQL grammar that I need to dig out and dust off that would help stretch this stuff a bit more.

**Comment by ejones** [comment]
Hey, wanted to drop in quick - love this work @ochafik! Have been following along the discussions from the notifications. Hope to have some time in the next few days to take a deeper look.

I took a super quick look and one surface level thought I had was: should the token decoding cache be on the `llama_context` (although I know it's already overloaded)? Decoding tokens is driven by the model vocabulary rather than anything about the grammar. And _technically_ I think as written the same grammar struct could be applied to different models.

Other than that, I need to have a think about UTF-8 decoding and that multibyte sequences with this.

**Comment by ochafik** [comment]
Thanks @ejones !

> should the token decoding cache be on the llama_context (although I know it's already overloaded)? Decoding tokens is driven by the model vocabulary rather than anything about the grammar.

Ahh, I wondered about this, guess I took the lazy route. If we moved this to the context we'd probably want to build it upfront to avoid concurrency issues in `server` arising from building the cache lazily. It would have to be done regardless of whether a grammar will later be used, which may be slightly wasteful, but then it would make using a new grammar marginally faster every time, so probably worth it. Will look into it!

> And technically I think as written the same grammar struct could be applied to different models.

Oh indeed, for speculative decoding I suppose the two models could have (ever so slightly) different tokenizers 😓

llama_context, hold on tight, more things coming your way :-D

**Comment by ochafik** [comment]
> should the token decoding cache be on the llama_context

@ejones done :-)

File: llama.cpp [hunk 2]
```diff
@@ -13615,21 +13624,26 @@ void llama_sample_grammar(struct llama_context * ctx, llama_token_data_array * c
         }
     }
 
+    // Store decoded codepoints when they are not cached (happens when there's a partial utf8 string prefix).
     std::vector<std::pair<std::vector<uint32_t>, llama_partial_utf8>> candidates_decoded;
-    candidates_decoded.reserve(candidates->size);
+    if (grammar->partial_utf8.n_remain > 0) {
+        candidates_decoded.reserve(candidates->size);
```
Comment (HanClinto) [inline llama.cpp:13630]: Almost feels like -- in the case of no partial_utf8 -- we could almost just copy the entire `ctx->token_codepoints_without_partial_utf8_prefix` array into `candidates_decoded` and call it a day -- is there a reason to rebuild this by calling `push_back` on every element individually?

Probably not a huge time-save, but just thinking out loud as I review this code again.

Comment (ochafik) [reply llama.cpp:13630]: Yes! And that's what is being done a few lines below in the `if (grammar->partial_utf8.n_remain == 0){` branch :-) In case of non partial_utf8 it skips the candidates_decoded step and just references the cached entries in the element pushed to candidates_grammar (no copies needed). Tried to comment this above but maybe it deserves a name change too, lemme know.

File: llama.cpp [hunk 3]
```diff
@@ -13037,6 +13042,10 @@ static std::vector<llama_grammar_candidate> llama_grammar_reject_candidates_for_
         }
     }
 
+    if (next_candidates.empty()) {
+        return rejects;
```
Comment (HanClinto) [inline llama.cpp:13046]: This is such a small and isolated change, I almost wonder if it shouldn't be pulled out into its own PR so that we can evaluate this performance improvement separate from the other one. As it is, it's difficult to know how much to attribute to each change...?

File: llama.cpp (hunk 2 above)
Comment (HanClinto) [reply llama.cpp:13630]: Oh whoops, I mixed up candidates_decoded with candidates_grammar.

I think what I was imagining was something like:

```
    std::vector<llama_grammar_candidate>                              candidates_grammar;
    // Store decoded codepoints when they are not cached (happens when there's a partial utf8 string prefix).
    std::vector<std::pair<std::vector<uint32_t>, llama_partial_utf8>> candidates_decoded;
    if (grammar->partial_utf8.n_remain > 0) {
        candidates_decoded.reserve(candidates->size);
        candidates_grammar.reserve(candidates->size);
    } else {
        // If there are no partial utf8 strings, we can use the cached codepoints directly.
        // Copy directly from ctx->token_codepoints_without_partial_utf8_prefix to candidates_grammar and ignore candidates_decoded.
        candidates_grammar.assign(ctx->token_codepoints_without_partial_utf8_prefix.begin(), ctx->token_codepoints_without_partial_utf8_prefix.end());
    }
```
And then I was imagining that one could skip the for-loop entirely (or at least greatly shorten it).

However, after reading through things again, that doesn't work for a few reasons -- not every token needs to be added to `candidates_grammar`.

Also, `token_codepoints_without_partial_utf8` is of type `std::vector<std::pair<std::vector<uint32_t>, llama_partial_utf8>>`, and `candidates_grammar` is of type `std::vector<llama_grammar_candidate>`, so the data types don't line up.

Though speaking of which, I wonder if the data type of `token_codepoints_without_partial_utf8` should be changed to match `candidates_grammar`? It would simplify the line a little lower to be:

```
        } else if (grammar->partial_utf8.n_remain == 0){
            const auto & decoded = ctx->token_codepoints_without_partial_utf8_prefix.at(id);
            candidates_grammar.push_back(decoded);
```

?

File: llama.cpp (hunk 3 above)
Comment (ochafik) [reply llama.cpp:13046]: It happens to be the [first commit](https://github.com/ggerganov/llama.cpp/pull/6811/commits/98f33bae767dd19e213ef663b22ad99979ca71d7) on the branch so you can git checkout before / after it and compare performance as follows:

```bash
( export COMMON_ARGS=(
    -mu https://huggingface.co/NousResearch/Hermes-2-Pro-Mistral-7B-GGUF/resolve/main/Hermes-2-Pro-Mistral-7B.Q4_K_M.gguf
    -m models/Hermes-2-Pro-Mistral-7B.Q4_K_M.gguf
    --prompt-cache issue4218.bin
    --grammar-file issue4218.gbnf
    -f issue4218.txt
    -c 3400
  ) && \
  hyperfine --warmup 1 --runs 5 \
    -L branch 98f33bae767dd19e213ef663b22ad99979ca71d7^,98f33bae767dd19e213ef663b22ad99979ca71d7 \
    --setup "\
      git checkout {branch} && \
      make clean && make -j LLAMA_CURL=1 main && \
      rm -f issue4218.bin && \
      ./main ${COMMON_ARGS[*]} -n 1" \
    "BRANCH={branch} \
      ./main ${COMMON_ARGS[*]} -n 128 --prompt-cache-ro --seed 12345 --no-display-prompt" )
```

<details>
<summary>show output</summary>

```
Benchmark 1: BRANCH=98f33bae767dd19e213ef663b22ad99979ca71d7^       ./main -mu https://huggingface.co/NousResearch/Hermes-2-Pro-Mistral-7B-GGUF/resolve/main/Hermes-2-Pro-Mistral-7B.Q4_K_M.gguf -m models/Hermes-2-Pro-Mistral-7B.Q4_K_M.gguf --prompt-cache issue4218.bin --grammar-file issue4218.gbnf -f issue4218.txt -c 3400 -n 128 --prompt-cache-ro --seed 12345 --no-display-prompt
  Time (mean ± σ):      7.970 s ±  0.060 s    [User: 3.829 s, System: 0.292 s]
  Range (min … max):    7.877 s …  8.025 s    5 runs
 
Benchmark 2: BRANCH=98f33bae767dd19e213ef663b22ad99979ca71d7       ./main -mu https://huggingface.co/NousResearch/Hermes-2-Pro-Mistral-7B-GGUF/resolve/main/Hermes-2-Pro-Mistral-7B.Q4_K_M.gguf -m models/Hermes-2-Pro-Mistral-7B.Q4_K_M.gguf --prompt-cache issue4218.bin --grammar-file issue4218.gbnf -f issue4218.txt -c 3400 -n 128 --prompt-cache-ro --seed 12345 --no-display-prompt
  Time (mean ± σ):      5.814 s ±  0.037 s    [User: 1.674 s, System: 0.277 s]
  Range (min … max):    5.758 s …  5.857 s    5 runs
 
Summary
  'BRANCH=98f33bae767dd19e213ef663b22ad99979ca71d7       ./main -mu https://huggingface.co/NousResearch/Hermes-2-Pro-Mistral-7B-GGUF/resolve/main/Hermes-2-Pro-Mistral-7B.Q4_K_M.gguf -m models/Hermes-2-Pro-Mistral-7B.Q4_K_M.gguf --prompt-cache issue4218.bin --grammar-file issue4218.gbnf -f issue4218.txt -c 3400 -n 128 --prompt-cache-ro --seed 12345 --no-display-prompt' ran
    1.37 ± 0.01 times faster than 'BRANCH=98f33bae767dd19e213ef663b22ad99979ca71d7^       ./main -mu https://huggingface.co/NousResearch/Hermes-2-Pro-Mistral-7B-GGUF/resolve/main/Hermes-2-Pro-Mistral-7B.Q4_K_M.gguf -m models/Hermes-2-Pro-Mistral-7B.Q4_K_M.gguf --prompt-cache issue4218.bin --grammar-file issue4218.gbnf -f issue4218.txt -c 3400 -n 128 --prompt-cache-ro --seed 12345 --no-display-prompt'
```

</details>

It doesn't help w/ all grammars, though.

File: llama.cpp (hunk 2 above)
Comment (HanClinto) [reply llama.cpp:13630]: > Though speaking of which, I wonder if the data type of token_codepoints_without_partial_utf8 should be changed to match candidates_grammar?

Answering my own question, it would simplify that particular line, but it would complicate the changes in this section:

```
    // Note terminating 0 in decoded string
    const auto   decoded     = grammar->partial_utf8.n_remain == 0
        ? ctx->token_codepoints_without_partial_utf8_prefix[token]
        : decode_utf8(piece, grammar->partial_utf8);
```

I think the way you have it here is fine -- don't worry about changing anything!

**Comment by HanClinto** [comment]
Overall this looks really good! Nice, clean work!

Do we have any reservations about increasing the memory footprint of the context? The caching code increases memory usage roughly by the vocabulary * 2, I think -- is that a problem?

The early exit in `llama_grammar_reject_candidates_for_stack` feels like a no-brainer easy win, but I don't know enough about the context and the memory usage goals to know if the caching is acceptable or not.

File: llama.cpp [hunk 4]
```diff
@@ -15714,6 +15730,18 @@ struct llama_context * llama_new_context_with_model(
         }
     }
 
+    // cache tokens & their decoded codepoints (for common case where there's no partial utf8 prefix bytes) for grammar-constrained sampling.
+    {
```
Comment (HanClinto) [inline llama.cpp:15734]: Only reservation that I have is that we're caching these data structures whether there is a grammar or not.  These data structures are only used in `llama_sample_grammar` and `llama_grammar_accept_token`, so if there is no grammar, then this is wasted time (and perhaps more importantly, memory).

Comment (ochafik) [reply llama.cpp:15734]: Fair point! A “grammars_enabled” context param might make sense, with a flag to turn it off in server (+ explode when grammar requested), and turned on in main only when grammar or schema flag set. Would save a couple of MBs and a tiny bit of preprocessing. Will add tonight 👌

Comment (HanClinto) [reply llama.cpp:15734]: Could also perhaps restructure this cache building that happens as a memorization step that happens inside of the first calls to decode_utf8 / llama_token_to_piece, rather than when the context is built.

Comment (ochafik) [reply llama.cpp:15734]: That’s what I initially did (and was storing it in the grammar itself, which was pointed out as a not ideal by @ejones ). Lazy init comes with big concurrency concerns (say two slots start working with grammars at the same time in the server: can’t let them both populate the data lazily if it’s in the context)

Comment (ochafik) [reply llama.cpp:15734]: (Oh but then, context is slot-specific? I’ll look again tonight 😅)

Comment (ochafik) [reply llama.cpp:15734]: > (Oh but then, no sure anymore whether context is slot-specific? I’ll look again tonight 😅)

Yeah context is shared. And as @ejones pointed out grammar itself technically could be shared in other contexts.

I played with the idea of allowing to disable the preprocessing ([here](https://github.com/ochafik/llama.cpp/compare/grammar-fast...grammar-fast-disable?expand=1)) but I haven't found a good way to articulate the flag wrt/ all the ways the API can be used. I think it'd be simpler to leave it as is and keep it as an area where to potentially squeeze a couple of MB when times are scarce. wdyt?

Comment (HanClinto) [reply llama.cpp:15734]: Apologies in advance for this comment -- this is very long, and very stream-of-consciousness, so please take everything with a grain of salt. However, I want to keep moving forward on this PR, so I figure at least _some_ reply is better than nothing -- even if it's a bit rambly. :)

> I played with the idea of allowing to disable the preprocessing ([here](https://github.com/ochafik/llama.cpp/compare/grammar-fast...grammar-fast-disable?expand=1)) but I haven't found a good way to articulate the flag wrt/ all the ways the API can be used.

Yeah, that's a tricky question. I wonder if instead of a command-line flag, if the precache step at the end of `llama_new_context_with_model ` should maybe be broken out into a separate call -- something like `llama_context_precache_grammar_token_pieces`? Then, whoever is initializing the context can also add precaching to it...? calling the API can enable this pre-caching feature if they want to or not, but it takes an extra call...?

So I'm back to wondering if there is a way where we can still break this precache functionality into a separate function. Can we check for the need of it in `llama_sample_grammar` / `llama_grammar_accept_token` and call the precaching there as-needed? 

```
if (ctx->token_pieces.empty()) {
    llama_context_precache_grammar_token_pieces(ctx);
}
const auto & piece = ctx->token_pieces.at(token);
...
```

It's annoying to run that check on every token, but the performance improvements when using a grammar should still win out. And we also remove the extra memory usage in the case of no grammar being used.

But as I'm thinking through this now, I think what you said earlier is the problem of ctx being shared across multiple threads / processes, so at the very least, we would need to add a blocking mutex here. Is that feasible / reasonable? We do a number of other little mutexes scattered throughout the code in other places, so it feels like it wouldn't be awful here...?

So yeah, I'm still on the fence about making it as a separate call that needs to be called manually if one wants the performance boost (maybe with an automatic fallback mechanism in the sampler), or else transparent caching on first call at runtime with a mutex (provided we can implement this without it being too annoying to run this check inside such a tight loop).

> I think it'd be simpler to leave it as is and keep it as an area where to potentially squeeze a couple of MB when times are scarce. wdyt?

This also sounds not unreasonable, but I don't know how to weigh such things. I know I really like grammar-constrained sampling, but I don't know how popular the feature is overall, and is it worth negatively impacting hyper-resource-constrained usages (such as Raspberry Pis or whatnot) vs. grammars? That's what I'm unable to weigh -- I feel like that's a strategic decision that's a bit above my level.

Comment (gabriel-peracio) [reply llama.cpp:15734]: > I don't know how popular the feature is overall

User here: I've been finding the feature extremely useful and can't live without it. There are also no substitutes anywhere, the competitor's solutions are worse than llama.cpp's implementation IMO

Comment (ochafik) [reply llama.cpp]: @HanClinto I've moved the caching back to llama_sample_grammar in lazy & mutex-protected form (thanks for the suggestion!). And sent the early exit change separately -> https://github.com/ggerganov/llama.cpp/pull/7370

**Comment by ochafik** [comment]
Confirming this is superseded by https://github.com/ggerganov/llama.cpp/pull/7587 & https://github.com/ggerganov/llama.cpp/pull/7424, no incremental gains from caching the decoded codepoints
